File: ruby/lib/runapi/suno/resources/create_mashup.rb
# frozen_string_literal: true

module RunApi
  module Suno
    module Resources
      # Blends two audio tracks into a single new composition.
      class CreateMashup
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/suno/create_mashup"
        RESPONSE_CLASS = Types::CreateMashupResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedCreateMashupResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          Validators.validate_create_mashup!(params, self)
        end
      end
    end
  end
end

File: python/src/runapi/suno/resources/create_mashup.py
"""Suno create_mashup resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource, ValidationError

from ..contract_gen import CONTRACT
from ..types import CompletedCreateMashupResponse, CreateMashupResponse


class CreateMashup(Resource):
    """Create a mashup from two uploaded tracks."""

    ENDPOINT = "/api/v1/suno/create_mashup"

    RESPONSE_CLASS = CreateMashupResponse
    COMPLETED_RESPONSE_CLASS = CompletedCreateMashupResponse

    def run(self, **params: Any) -> Any:
        """Blend two tracks into one and poll until it completes.

        Args:
            **params: mashup parameters.

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create a mashup task and return immediately with an id.

        Args:
            **params: mashup parameters.

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of a mashup task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        upload_url_list = params.get("upload_url_list")
        if not (isinstance(upload_url_list, list) and len(upload_url_list) == 2):
            raise ValidationError("upload_url_list must contain exactly 2 URLs")
        self._validate_contract(CONTRACT["create-mashup"], params)
